### PR TITLE
Zen2: Add basic Zen1 transport-level BWC

### PR DESCRIPTION
**Pull request by ywelsch**
Implements serialization compatibility between Zen1 and Zen2 transport action, allowing a Zen1 node join a Zen2 master and vice-versa.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
I left a few nits. It looks like we're missing handlers for `internal:discovery/zen/rejoin` and `internal:discovery/zen/leave`. Not sure we need to make them do anything interesting here yet, but it'd be good to have placeholders so they don't get forgotten.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 1]
```diff
@@ -1110,7 +1111,15 @@ protected void sendPublishRequest(DiscoveryNode destination, PublishRequest publ
         @Override
         protected void sendApplyCommit(DiscoveryNode destination, ApplyCommitRequest applyCommit,
                                        ActionListener<Empty> responseActionListener) {
-            publicationHandler.sendApplyCommit(destination, applyCommit, wrapWithMutex(responseActionListener));
+            publicationContext.sendApplyCommit(destination, applyCommit, wrapWithMutex(responseActionListener));
         }
     }
+
+    public static Settings.Builder addZen1Attribute(Settings.Builder builder) {
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java]: added in 9b12b05d241

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 2]
```diff
@@ -1110,7 +1111,15 @@ protected void sendPublishRequest(DiscoveryNode destination, PublishRequest publ
...
+        return builder.put("node.attr.zen1", true);
+    }
+
+    public static boolean isZen1Node(DiscoveryNode discoveryNode) {
+        return discoveryNode.getVersion().before(Version.V_7_0_0) || discoveryNode.getAttributes().containsKey("zen1");
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java]: added in 9b12b05d241

File: server/src/main/java/org/elasticsearch/cluster/coordination/FollowersChecker.java [hunk 3]
```diff
@@ -290,7 +299,18 @@ private void handleWakeUp() {
...
             final FollowerCheckRequest request = new FollowerCheckRequest(fastResponseState.term, transportService.getLocalNode());
             logger.trace("handleWakeUp: checking {} with {}", discoveryNode, request);
-            transportService.sendRequest(discoveryNode, FOLLOWER_CHECK_ACTION_NAME, request,
+
+            final String actionName;
+            final TransportRequest transportRequest;
+            if (Coordinator.isZen1Node(discoveryNode)) {
+                actionName = NodesFaultDetection.PING_ACTION_NAME;
+                transportRequest = new NodesFaultDetection.PingRequest(discoveryNode, ClusterName.CLUSTER_NAME_SETTING.get(settings),
+                    transportService.getLocalNode(), 0L);
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/FollowersChecker.java]: fixed in cf3d83e7493

File: server/src/main/java/org/elasticsearch/discovery/PeerFinder.java [hunk 4]
```diff
@@ -384,50 +401,76 @@ private void requestPeers() {
...
+                    return Names.GENERIC;
+                }
+            };
+            final String actionName;
+            final TransportRequest transportRequest;
+            final TransportResponseHandler<?> transportResponseHandler;
+            if (Coordinator.isZen1Node(discoveryNode)) {
+                actionName = UnicastZenPing.ACTION_NAME;
+                transportRequest = new UnicastZenPing.UnicastPingRequest(1, ZenDiscovery.PING_TIMEOUT_SETTING.get(settings),
+                    new ZenPing.PingResponse(getLocalNode(), null, ClusterName.CLUSTER_NAME_SETTING.get(settings), 0L));
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/discovery/PeerFinder.java]: fixed in cf3d83e7493

File: server/src/main/java/org/elasticsearch/discovery/PeerFinder.java [hunk 5]
```diff
@@ -439,4 +482,19 @@ public String toString() {
...
+    private class Zen1UnicastPingRequestHandler implements TransportRequestHandler<UnicastZenPing.UnicastPingRequest> {
+        @Override
+        public void messageReceived(UnicastZenPing.UnicastPingRequest request, TransportChannel channel, Task task) throws Exception {
+            final PeersRequest peersRequest = new PeersRequest(request.pingResponse.node(),
+                Optional.ofNullable(request.pingResponse.master()).map(Collections::singletonList).orElse(Collections.emptyList()));
+            final PeersResponse peersResponse = handlePeersRequest(peersRequest);
+            final List<ZenPing.PingResponse> pingResponses = new ArrayList<>();
+            final ClusterName clusterName = ClusterName.CLUSTER_NAME_SETTING.get(settings);
+            pingResponses.add(new ZenPing.PingResponse(transportService.getLocalNode(), peersResponse.getMasterNode().orElse(null),
+                clusterName, 0L));
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/discovery/PeerFinder.java]: fixed in cf3d83e7493

File: server/src/main/java/org/elasticsearch/cluster/coordination/FollowersChecker.java [hunk 6]
```diff
@@ -103,8 +108,12 @@ public FollowersChecker(Settings settings, TransportService transportService,
         followerCheckRetryCount = FOLLOWER_CHECK_RETRY_COUNT_SETTING.get(settings);
 
         updateFastResponseState(0, Mode.CANDIDATE);
-        transportService.registerRequestHandler(FOLLOWER_CHECK_ACTION_NAME, Names.SAME, FollowerCheckRequest::new,
+        transportService.registerRequestHandler(FOLLOWER_CHECK_ACTION_NAME, Names.SAME, false, false, FollowerCheckRequest::new,
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/FollowersChecker.java]: That's a good question. I wonder if it's something we should discuss with a larger group, but go with Zen1 defaults for now.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**